sql: move function properties to overload level

chengxiong-ruan opened the pull request:
There're 3 commits:
(1) sql: use Overload in FunctionDefinition instead of overloadImpl

It feels a bit annoying to always need to cast overloadImpl
to Overload. In fact, all `overloadImpl`s in FunctionDefinition
are `Overload`s

(2) sql: move function properties to overload level

Moving function propterties to overload level, but still keep
it in FunctionDefiniton since it's still needed for internal
usage like docgen. Later on, the function resolution interface
will return a resolved version of function definition which
won't have the properties field. Instead, we'll go through
overloads to fetch properties for resolved function definition.
Current FunctionDefiniton will be kept for builtin functions only.

(3) sql: remove usage of GetBuiltinProperties from GetSequenceFromFunc

In `GetSequenceFromFunc` we resolve a functiong and then use the
name to get builtin properties. This seems unnecessary because
we may just get function properties from the resolved function
definition. With this change, we are eleminating a good amount of
usage cases of "GetBuiltinProperties", which seems to be the
majority of the usages of "GetBuiltinProperties".

There shouldn't be any functionality changes since we don't have UDFs yet.

chengxiong-ruan committed:
sql: use Overload in FunctionDefinition instead of overloadImpl

It feels a bit annoying to always need to cast overloadImpl
to Overload. In fact, all `overloadImpl`s in FunctionDefinition
are `Overload`s

Release note: None

chengxiong-ruan committed:
sql: move function properties to overload level

Moving function properties to overload level, but still keep
it in FunctionDefiniton since it's still needed for internal
usage like docgen. Later on, the function resolution interface
will return a resolved version of function definition which
won't have the properties field. Instead, we'll go through
overloads to fetch properties for resolved function definition.
Current FunctionDefiniton will be kept for builtin functions
for now until we have another refactoring to deprecate it. This
can be done when we have the function resolver implemented.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85199)
<!-- Reviewable:end -->

chengxiong-ruan commented on pkg/sql/sem/tree/type_check.go:
```diff
@@ -1047,10 +1043,17 @@ func (expr *FuncExpr) TypeCheck(
 		return nil, err
 	}
 
-	if err := semaCtx.checkFunctionUsage(expr, def); err != nil {
```
This check is moved to after overload resolution is done

chengxiong-ruan committed:
sql: remove usage of GetBuiltinProperties from GetSequenceFromFunc

In `GetSequenceFromFunc` we resolve a function and then use the
name to get builtin properties. This seems unnecessary because
we may just get function properties from the resolved function
definition. With this change, we are eleminating a good amount of
usage cases of "GetBuiltinProperties", which seems to be the
majority of the usages of "GetBuiltinProperties".

Release note: None

ajwerner reviewed (commented):
> Reviewed 9 of 9 files at r1, 9 of 9 files at r2, 19 of 19 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @mgartner)

___
*[`-- commits` line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85h1_K8bFq5wsvCYFN):*
nit: `functiong`
___
*[`pkg/sql/opt/optbuilder/groupby.go` line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85f27Y00Nc8UwZ0Qzq:-N85f27Y00Nc8UwZ0Qzr:b3746xy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/groupby.go#L883)):*
> ```Go
> 		return false, err
> 	}
> 	return cls == tree.AggregateClass, nil
> ```

make yourself a helper?

```go
func isClass(def *tree.FunctionDefinition, want tree.FunctionClass) (bool, error) { ... }
```

Also, you're in `optbuilder` so I think the helper could just panic and make the code easier to use?



_Code quote:_
```Go
	cls, err := def.GetClass()
	if err != nil {
		return false, err
	}
	return cls == tree.AggregateClass, nil
```
___
*[`pkg/sql/opt/optbuilder/srfs.go` line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85gCFDBl2JKUHi26U3:-N85gCFDBl2JKUHi26U4:b-nfwdbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/srfs.go#L117)):*
> ```Go
> 			}
> 			if funcCls != tree.GeneratorClass || b.shouldCreateDefaultColumn(texpr) {
> 				returnLabels, err := def.GetReturnLabel()
> ```

this one will be fun one day. I have to imagine we can have different return labels, no? For another day. 
___
*[`pkg/sql/sem/transform/aggregates.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85gXR5EYlwHRbdpvJW:-N85gXR5EYlwHRbdpvJX:bevvakd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/transform/aggregates.go#L44)):*
> ```Go
> 		funcCls, err := fd.GetClass()
> 		if err != nil {
> 			panic(err)
> ```

I don't think this is kosher. I also don't think this `IsAggregateVisitor` should be exported. Instead, I think we should 1) unexport it and 2) follow the pattern used by the `normalize.Visitor` where we stash the error internally. 
___
*[`pkg/sql/sem/tree/function_definition.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85do2h7YTQDLYFlXX_:-N85do2h7YTQDLYFlXXa:b-410wnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L29)):*
> ```Go
> 
> 	// FunctionProperties are the properties common to all overloads.
> 	FunctionProperties
> ```

Can you add commentary here about the intention to remove this. Honestly, if you're willing, can you un-embed it and name it `DeprecatedFunctionProperties`, or if that's too much, can you make a new type `type DeprecatedFunctionProperties struct { FunctionProperties }` and embed that?



_Code quote:_
```Go
	// FunctionProperties are the properties common to all overloads.
	FunctionProperties
```
___
*[`pkg/sql/sem/tree/type_check.go` line 948 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85eOxk5lwO3dda67kz:-N85eOxk5lwO3dda67l-:b8yut3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/type_check.go#L948)):*
> ```Go
> 	}
> 
> 	// TODO(Chengxiong): Consider doing this check when we narrowing down to an
> ```

grammar nit: `when we are narrowing` or `when we narrow`
___
*[`pkg/sql/sem/tree/type_check.go` line 1177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85ejjK6cFZCeyQkg4n:-N85ejjK6cFZCeyQkg4o:b9tbm17) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/type_check.go#L1177)):*
> ```Go
> 		return nil, pgerror.Wrapf(errPrivateFunction, pgcode.ReservedName,
> 			"%s()", errors.Safe(def.Name))
> 	}
> ```

this is new, why are you adding it? 



_Code quote:_
```Go
	if overloadImpl.Private {
		return nil, pgerror.Wrapf(errPrivateFunction, pgcode.ReservedName,
			"%s()", errors.Safe(def.Name))
	}
```


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
I don't love some of the hacks in the second commit. Keeping properties in FunctionDefinition will confuse us and be error prone. And I'm hesistant about adding the ambiguous overload errors before we've added tests for them and defined what the behavior. But if these things are temporary, that's fine. I made some suggestions for TODOs for things which I'm assuming we are intending to change in the near future.

> Reviewed 9 of 9 files at r1, 9 of 9 files at r2, 19 of 19 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-:-N85kjA5C9Y1IIqm_Kd2:bo0gtw7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`-- commits` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85fGdB5dt9P39hfPuu):*
nit: propterties => properties
___
*[`-- commits` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85f9Lp3K6r3hM2CMJD):*
Can you explain why it would need to exist at all? Seems like we could get rid of it entirely.
___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85fF3c0LHDHGRQ5HsJ):*
nit: remove `.`
___
*[`pkg/sql/opt/optbuilder/groupby.go` line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85fVZT1MHPMf1nOrpH:-N85fVZT1MHPMf1nOrpI:bjf6mwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/groupby.go#L883)):*
> ```Go
> 		return false, err
> 	}
> 	return cls == tree.AggregateClass, nil
> ```

nit: all these functions can be simplified to

```
cls, err := def.GetClass()
return cls == tree.SomeClass, err
```

I think it's commonly understood that if an error is returned from the function, the other return values are undefined. So there is no harm if `true` is returned here.
___
*[`pkg/sql/opt/optbuilder/scalar.go` line 543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85fzNE6uh6bJn5xL0a:-N85fzNE6uh6bJn5xL0b:b-da4kx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scalar.go#L543)):*
> ```Go
> 	}
> 
> 	if isAgg, err := isAggregate(def); err != nil {
> ```

At this point, we've already resolved the overload - notice `f.ResolvedOverload()` above. Therefore the overload class should never be ambiguous. Can we avoid this messy error code by checking the class of the overload instead?
___
*[`pkg/sql/opt/optbuilder/scope.go` line 1025 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85hPdY6jsJWJD7P6Vt:-N85hPdZBYbXjcYBu216:bjs6638) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scope.go#L1025)):*
> ```Go
> 		if isGen, err := isGenerator(def); err != nil {
> 			panic(err)
> 		} else if isGen && s.replaceSRFs {
> ```

Unfortunately, it doesn't look like the function has not been type-checked at this point, so the overload has not been resolved. This will probably have to change to support function overloads with the same name but different classes. We'll get ambiguous errors otherwise. Can you leave a TODO for me here so I can look into re-working this?
___
*[`pkg/sql/opt/optbuilder/scope.go` line 1144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85j1xg-BZ24UsZWsGg:-N85j1xg-BZ24UsZWsGh:b-e906sj) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scope.go#L1144)):*
> ```Go
> 
> 	// We moved function properties to overload level as well for better UDF
> 	// resolution. So we need to mark private=false on each overload as well.
> ```

nit: a comment saying that we moved something reads funny because its temporal. Can you just write "FunctionProperties exist in function definitions and their overloads, so we unset all private fields here. 
___
*[`pkg/sql/opt/optbuilder/scope.go` line 1145 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85ivGS9WNzqFgNrLOo:-N85ivGS9WNzqFgNrLOp:b-kepi5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scope.go#L1145)):*
> ```Go
> 	// We moved function properties to overload level as well for better UDF
> 	// resolution. So we need to mark private=false on each overload as well.
> 	unsetPrivate := func(overloads []*tree.Overload) {
> ```

nit: make this function take a def and set the definition and overload Private to false.
___
*[`pkg/sql/opt/optbuilder/srfs.go` line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85jhPoBKQdkS4gSFlR:-N85jhPoBKQdkS4gSFlS:b-ai7hs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/srfs.go#L114)):*
> ```Go
> 			funcCls, err := def.GetClass()
> 			if err != nil {
> 				panic(err)
> ```

We've already type checked the function at this point, so we should have a single overload. Can we just use the definition in that overload?
___
*[`pkg/sql/sem/tree/function_definition.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85aj1r9kh6S66OiD5U:-N85aj1r9kh6S66OiD5V:bw6wrcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a671ea2c43dc3bf94bb4505e2c343a0096ad2e03/pkg/sql/sem/tree/function_definition.go#L22)):*
> ```Go
> 
> 	// Definition is the set of overloads for this function name.
> 	Definition []*Overload
> ```

Why not `[]Overload`? If you do make it that, you might want to change some of the loops over Definition to avoid copying the overload by value.
___
*[`pkg/sql/sem/tree/function_definition.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85d5JDD9EK7aTprezf:-N85d5JDD9EK7aTprezg:b-vmyjnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L29)):*
> ```Go
> 
> 	// FunctionProperties are the properties common to all overloads.
> 	FunctionProperties
> ```

nit: leave a TODO to remove this now that it is obsolete
___
*[`pkg/sql/sem/tree/function_definition.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85egPV8JRWZczr1g0u:-N85egPV8JRWZczr1g0v:bb62md9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L186)):*
> ```Go
> 
> // GetClass returns function class by going through all overloads. Error is
> // return if it's ambiguous.
> ```

nit: for this function and the ones below, make the comment more specific than "going through all overloads", e.g. what happens when there is an overload with different classes. 
___
*[`pkg/sql/sem/tree/function_definition.go` line 187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85dfhS7ku_Ae79l9aU:-N85dfhS7ku_Ae79l9aV:b842rzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L187)):*
> ```Go
> // GetClass returns function class by going through all overloads. Error is
> // return if it's ambiguous.
> func (fd *FunctionDefinition) GetClass() (FunctionClass, error) {
> ```

Are these functions temporary? If not, this seems like the wrong direction. If class, return label, etc. are attributes of an overload, then this function shouldn't exist - all places that we use it should be modified to either resolve the overload before fetching a function's class/etc, or using the class/etc of multiple overloads to aid in resolution. 

I think in the long term we should get rid of `FunctionDefinition` entirely. The hierarchy of FunctionDefintions with many Overloads doesn't seem to match the concept we are trying to represent - it should be flattened to simply Overloads.
___
*[`pkg/sql/sem/tree/overload.go` line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85dGO41KWljuwRAlEY:-N85dGO41KWljuwRAlEZ:b-dqb4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/overload.go#L176)):*
> ```Go
> 	NullableArgs bool
> 
> 	FunctionProperties
> ```

nit: add a comment


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
Thanks for the reviews. And I feel you. And this is what made me sad for a while :( the ambiguous thing is needed because we can guarantee that a func expression is type checked already at each use case of function resolution. To get rid of it we need to make sure everything is type checked or it's ok with ambiguity in some cases. But just want to let you know that I'm definitely targeting to the direct you want as well, a flat level of overloads!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-:-N866JOY22FFPByvRZM2:bxpa39b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[`-- commits` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85f9Lp3K6r3hM2CMJD):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you explain why it would need to exist at all? Seems like we could get rid of it entirely.
</blockquote></details>

It doesn't need to exist. We can git rid of it when we have the function resolver implemented and replace it. Although, there're still cases we need to access builtin function properties, those use cases can be refactored to use `GetBuiltinProperties`. I added more words to the commit message. I'll also add TODOs for future refactorings.
___
*[`pkg/sql/opt/optbuilder/groupby.go` line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85f27Y00Nc8UwZ0Qzq:-N85s8vR9GwMukxOcYKZ:b-bzdkfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/groupby.go#L883)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

make yourself a helper?

```go
func isClass(def *tree.FunctionDefinition, want tree.FunctionClass) (bool, error) { ... }
```

Also, you're in `optbuilder` so I think the helper could just panic and make the code easier to use?
</blockquote></details>

yeah...panic within `isClass` is definitely cleaner.
___
*[`pkg/sql/opt/optbuilder/groupby.go` line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85fVZT1MHPMf1nOrpH:-N85sD_zD-0BcRoMLlvJ:b6bnzig) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/groupby.go#L883)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: all these functions can be simplified to

```
cls, err := def.GetClass()
return cls == tree.SomeClass, err
```

I think it's commonly understood that if an error is returned from the function, the other return values are undefined. So there is no harm if `true` is returned here.
</blockquote></details>

changed to panic within `isClass`, good suggestion as well though.
___
*[`pkg/sql/opt/optbuilder/scalar.go` line 543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85fzNE6uh6bJn5xL0a:-N85tR3HDlvUuBPXhgt2:bik6uhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scalar.go#L543)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

At this point, we've already resolved the overload - notice `f.ResolvedOverload()` above. Therefore the overload class should never be ambiguous. Can we avoid this messy error code by checking the class of the overload instead?
</blockquote></details>

That is totally true! changed to just checking the class of the resolved overload.
___
*[`pkg/sql/opt/optbuilder/scope.go` line 1025 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85hPdY6jsJWJD7P6Vt:-N85ulorAfYgsos4WzXU:b-m7uavw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scope.go#L1025)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Unfortunately, it doesn't look like the function has not been type-checked at this point, so the overload has not been resolved. This will probably have to change to support function overloads with the same name but different classes. We'll get ambiguous errors otherwise. Can you leave a TODO for me here so I can look into re-working this?
</blockquote></details>

Yeah, Added a TODO. One thing we could also do is to utilized the inputs of the function expression to narrow the scope of overloads a bit. I also leave it as a TODO for myself at the function resolver interface definition.
___
*[`pkg/sql/opt/optbuilder/scope.go` line 1144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85j1xg-BZ24UsZWsGg:-N85vr9qD2H46Wb2tjLq:baqy283) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scope.go#L1144)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: a comment saying that we moved something reads funny because its temporal. Can you just write "FunctionProperties exist in function definitions and their overloads, so we unset all private fields here. 
</blockquote></details>

heh, totally true. fixed.
___
*[`pkg/sql/opt/optbuilder/scope.go` line 1145 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85ivGS9WNzqFgNrLOo:-N85wrLO3ICEIR-O7fQP:b4d84sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/scope.go#L1145)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: make this function take a def and set the definition and overload Private to false.
</blockquote></details>

good suggestions.
___
*[`pkg/sql/opt/optbuilder/srfs.go` line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85jhPoBKQdkS4gSFlR:-N85y_Ps2zV3zpmkMNeA:beqxvei) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/srfs.go#L114)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We've already type checked the function at this point, so we should have a single overload. Can we just use the definition in that overload?
</blockquote></details>

AH, didn't know that. I trust you and changed as suggested.
___
*[`pkg/sql/opt/optbuilder/srfs.go` line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85gCFDBl2JKUHi26U3:-N85yoW07lrJ1M5_RyLu:b4jobxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/opt/optbuilder/srfs.go#L117)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this one will be fun one day. I have to imagine we can have different return labels, no? For another day. 
</blockquote></details>

¯\_(ツ)_/¯
___
*[`pkg/sql/sem/transform/aggregates.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85gXR5EYlwHRbdpvJW:-N86-CbzDfV0-2mnnxkF:b-hwells) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/transform/aggregates.go#L44)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think this is kosher. I also don't think this `IsAggregateVisitor` should be exported. Instead, I think we should 1) unexport it and 2) follow the pattern used by the `normalize.Visitor` where we stash the error internally. 
</blockquote></details>

interesting. looks like it's fine to just return a `false, expr` if there's any error, similar to how it handles the resolution error a few lines above.
___
*[`pkg/sql/sem/tree/function_definition.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85aj1r9kh6S66OiD5U:-N86-Xye4mvU9PqLSWpr:b9zhghz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a671ea2c43dc3bf94bb4505e2c343a0096ad2e03/pkg/sql/sem/tree/function_definition.go#L22)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why not `[]Overload`? If you do make it that, you might want to change some of the loops over Definition to avoid copying the overload by value.
</blockquote></details>

In `NewFunctionDefinition` it took the pointer of `Overload`.  I actually don't know which one is better, to avoid copying overloads or to trust that nothing else change anything of it at a random place..
___
*[`pkg/sql/sem/tree/function_definition.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85do2h7YTQDLYFlXX_:-N860uNG5mh4ImAEzE_d:bcmtt5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L29)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you add commentary here about the intention to remove this. Honestly, if you're willing, can you un-embed it and name it `DeprecatedFunctionProperties`, or if that's too much, can you make a new type `type DeprecatedFunctionProperties struct { FunctionProperties }` and embed that?
</blockquote></details>

I added a TODO to remove this struct entirely. I'm visioning (actually some code WIP) to have a "resolved version" of something similar, where there is no function property, and also with each overload annotated with a schema name.

Also good suggestion on the wrapping thing. I actually thought about the same, but we just constructed `FunctionProperties` by doing something like `tree.FunctionProperties{Category: builtinconstants.CategoryString}` at a few hundreds of places, and I hesitated to fixe them all in this pr. With the TODO, I'll leave as it is.
___
*[`pkg/sql/sem/tree/function_definition.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85d5JDD9EK7aTprezf:-N861PR66L8Q78i-RxZZ:b-qwduw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L29)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: leave a TODO to remove this now that it is obsolete
</blockquote></details>

added. thanks.
___
*[`pkg/sql/sem/tree/function_definition.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85egPV8JRWZczr1g0u:-N865KiJ4RCAwOcfho4u:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L186)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: for this function and the ones below, make the comment more specific than "going through all overloads", e.g. what happens when there is an overload with different classes. 
</blockquote></details>

done.
___
*[`pkg/sql/sem/tree/function_definition.go` line 187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85dfhS7ku_Ae79l9aU:-N864I3n3VS50QKI5Cy9:b9xo38h) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/function_definition.go#L187)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Are these functions temporary? If not, this seems like the wrong direction. If class, return label, etc. are attributes of an overload, then this function shouldn't exist - all places that we use it should be modified to either resolve the overload before fetching a function's class/etc, or using the class/etc of multiple overloads to aid in resolution. 

I think in the long term we should get rid of `FunctionDefinition` entirely. The hierarchy of FunctionDefintions with many Overloads doesn't seem to match the concept we are trying to represent - it should be flattened to simply Overloads.
</blockquote></details>

Totally agree! I'm visioning to get rid of this `FunctionDefinition` thing. In short term, we may just have the function resolution return a list of overloads. However, we still need similar functions like these to check ambiguity. These can only be thrown away once we can make sure all the use cases of function resolution can get down to a single overload. I had this problem over my head for a while ( see this [old slack thread](https://cockroachlabs.slack.com/archives/C036BQTUJ9Z/p1656356651381859)). And I can't wait to get rid of it. But I think the scope of this work is too much for the short term goal because I'm not familiar with a lot of the use cases and how to make sure all of them can resolve to one overload.
___
*[`pkg/sql/sem/tree/overload.go` line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85dGO41KWljuwRAlEY:-N865W4LDSxBqCMT_3Z7:b-t948zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/overload.go#L176)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a comment
</blockquote></details>

added.
___
*[`pkg/sql/sem/tree/type_check.go` line 948 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85eOxk5lwO3dda67kz:-N865hx_6bLcQLApi7Da:b-o496ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/type_check.go#L948)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

grammar nit: `when we are narrowing` or `when we narrow`
</blockquote></details>

fixed. thanks.
___
*[`pkg/sql/sem/tree/type_check.go` line 1177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N85ejjK6cFZCeyQkg4n:-N865pCL1io9NQm4F4cM:bvb4h52) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd764fea087ca84b2b13c1a84fd92ca31418139b/pkg/sql/sem/tree/type_check.go#L1177)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is new, why are you adding it? 
</blockquote></details>

I actually moved this out from the `checkFunctionUsage` function. I think it's a good idea to do it here since we know the one best overload. I leave other properties check in `checkFunctionUsage` since they're much more complicated...


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
:+1: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 28 of 28 files at r4, 19 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-:-N86BTNr8b0kQNjpDVOi:bfrdjtf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @mgartner)

___
*[`pkg/sql/opt/optbuilder/scope.go` line 1141 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86C4ca3EKgIfKAz9so:-N86C4ca3EKgIfKAz9sp:b-dq1wai) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/opt/optbuilder/scope.go#L1141)):*
> ```Go
> 
> 	// FunctionProperties exist in function definitions and their overloads, so we
> 	// unset all private fields here
> ```

nit: add `.` at the end
___
*[`pkg/sql/sem/tree/function_definition.go` line 188 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86EuLa1sbpJxV1G6AP:-N86EuLa1sbpJxV1G6AQ:b-2tdvdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/sem/tree/function_definition.go#L188)):*
> ```Go
> }
> 
> // GetClass returns function class by checking each overload's Class and return
> ```

nit: return => returns
___
*[`pkg/sql/sem/tree/function_definition.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86F-9g7_PdJBHmW18g:-N86F-9g7_PdJBHmW18h:b-7a6cp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/sem/tree/function_definition.go#L198)):*
> ```Go
> }
> 
> // GetReturnLabel returns function ReturnLabel by checking each overload's and
> ```

nit: overload's => overload
___
*[`pkg/sql/sem/tree/function_definition.go` line 199 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86Exra75D3DIWmBkZu:-N86Exra75D3DIWmBkZv:bu6crle) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/sem/tree/function_definition.go#L199)):*
> ```Go
> 
> // GetReturnLabel returns function ReturnLabel by checking each overload's and
> // return a ReturnLabel if all overloads have ReturnLabel of same length.
> ```

nit: return => returns
nit: **a** ReturnLabel of **the** same length


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85199)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[`pkg/sql/opt/optbuilder/scope.go` line 1141 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86C4ca3EKgIfKAz9so:-N86Lzvo3gSpDincNFtJ:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/opt/optbuilder/scope.go#L1141)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add `.` at the end
</blockquote></details>

fixed.
___
*[`pkg/sql/sem/tree/function_definition.go` line 188 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86EuLa1sbpJxV1G6AP:-N86MCZB-KaPrxJ0XqgX:b-32hugm) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/sem/tree/function_definition.go#L188)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: return => returns
</blockquote></details>

fixed. thanks
___
*[`pkg/sql/sem/tree/function_definition.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86F-9g7_PdJBHmW18g:-N86MLEu1gEXH-VoBqU7:b-32hugm) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/sem/tree/function_definition.go#L198)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: overload's => overload
</blockquote></details>

fixed. thanks
___
*[`pkg/sql/sem/tree/function_definition.go` line 199 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85199#-N86Exra75D3DIWmBkZu:-N86MV10-_-fTjn0_WZD:b-t6aalr) ([raw file](https://github.com/cockroachdb/cockroach/blob/00048377ff4d847af7424e234832c92b0ca16944/pkg/sql/sem/tree/function_definition.go#L199)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: return => returns
nit: **a** ReturnLabel of **the** same length
</blockquote></details>

thanks. I hope I can do better on grammar things in the future.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
LGTM

chengxiong-ruan commented:
TFTR!
bors r+
